pkg/metastore/ms_client.go

```diff
@@ -0,0 +1,275 @@
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func Copy(ctx context.Context, fromTable *Table, partitions []*Partition, toDB, toTable, serde string, toClient WriteClient, transformLocation func(location string) (string, error)) error {
+	err := fromTable.Update(toDB, toTable, serde, transformLocation)
+	if err != nil {
+		return err
+	}
+	for _, partition := range partitions {
+		err := partition.Update(toDB, toTable, serde, transformLocation)
+		if err != nil {
+			return err
+		}
+	}
+	err = toClient.CreateTable(ctx, fromTable)
+	if err != nil {
+		return err
+	}
+	err = toClient.AddPartitions(ctx, toTable, toDB, partitions)
+	return err
+}
+
+func Merge(ctx context.Context, table *Table, partitionIter Collection, toDB, toTable, serde string, toClient Client, transformLocation func(location string) (string, error)) error {
+	err := table.Update(toDB, toTable, serde, transformLocation)
+	if err != nil {
+		return err
+	}
+	toPartitions, err := toClient.GetPartitions(ctx, toDB, toTable)
+	if err != nil {
+		return err
+	}
+	toPartitionIter := NewPartitionCollection(toPartitions)
+	var addPartitions, removePartitions, alterPartitions []*Partition
+	err = DiffIterable(partitionIter, toPartitionIter, func(difference catalog.DifferenceType, value interface{}, _ string) error {
+		partition, ok := value.(*Partition)
+		if !ok {
+			return fmt.Errorf("%w in diffIterable call. expected to get * Partition, but got: %T", ErrExpectedType, value)
+		}
+		err = partition.Update(toDB, toTable, serde, transformLocation)
+		if err != nil {
+			return err
+		}
+		switch difference {
+		case catalog.DifferenceTypeRemoved:
+			removePartitions = append(removePartitions, partition)
+		case catalog.DifferenceTypeAdded:
+			addPartitions = append(addPartitions, partition)
+		default:
+			alterPartitions = append(alterPartitions, partition)
+		}
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	err = toClient.AlterTable(ctx, toDB, toTable, table)
+	if err != nil {
+		return err
+	}
+	err = toClient.AddPartitions(ctx, toTable, toDB, addPartitions)
+	if err != nil {
+		return err
+	}
+	err = toClient.AlterPartitions(ctx, toDB, toTable, alterPartitions)
+	if err != nil {
+		return err
+	}
+	// drop one by one
+	for _, partition := range removePartitions {
+		values := partition.Values
+		err = toClient.DropPartition(ctx, toDB, toTable, values)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func CopyPartition(ctx context.Context, fromClient ReadClient, toClient Client, fromDB, fromTable, toDB, toTable, serde string, partition []string, transformLocation func(location string) (string, error)) error {
+	p1, err := fromClient.GetPartition(ctx, fromDB, fromTable, partition)
+	if err != nil {
+		return err
+	}
+	p2, err := toClient.GetPartition(ctx, toDB, toTable, partition)
+	if err != nil {
+		return err
+	}
+	err = p1.Update(toDB, toTable, serde, transformLocation)
+	if err != nil {
+		return err
+	}
```

Comment:
strange that copy operation perform update call on the source

Reply:
@nopcoder 
It updates the partition returned from the source and saves it to the destination.
Would changing the name of the function make it more clear?
Or maybe returning a new partition instead of changing it?